### ISSUE TITLE
Dune doesn't generate correct .merlin file when directory path contains space

### ISSUE THREAD

**Issue by HavinLeung** [ocaml/dune#4111]
## Expected Behavior

suppose I have a project as follows:
/foo/bar baz/dune-project
/foo/bar baz/lib/dune
/foo/bar baz/lib/foo.ml

if I run `dune build` from /foo/bar baz/, it generates /foo/bar baz/lib/.merlin

The 2nd last line in my merlin file should (probably) read:
`FLG -ppx ''\''/foo/bar\ baz/_build/default/.ppx/57ef......61/ppx.exe'\'' --as-ppx --cookie '\''library-name="my_lib"'\'''`

## Actual Behavior

The 2nd last line in my merlin file reads:
`FLG -ppx ''\''/foo/bar baz/_build/default/.ppx/57ef......61/ppx.exe'\'' --as-ppx --cookie '\''library-name="my_lib"'\'''`

Note that the space in "bar baz" isn't escaped. This causes issue with my editor (VScode with OCaml Platform and ocamllsp) not recognizing modules like Core or my ppx extensions.

## Reproduction

I believe I described in sufficient detail in expected behaviour

## Specifications

- Version of `dune` (output of `dune --version`): 2.7.1
- Version of `ocaml` (output of `ocamlc --version`): 4.07.1
- Operating system (distribution and version): MacOS Catalina 10.15.7 

## Additional information 

- running on a 16-inch macbook pro 2019 on a i7 processor (not the new ARM chips)
- using VSCode (1.52.1) with OCaml Platform (1.5.1) and ocamllsp (1.4.0)
- Link to gist with verbose output (run `dune` with the `--verbose` flag): https://gist.github.com/HavinLeung/6db2c575a3aa9671195533799834f045

**Comment by voodoos** [comment]
Thanks for your report.

Note that Dune 2.8.0, released today, completely reworks Dune / Merlin interactions.

However I experimented a bit and I am under the impression that this issue is still valid. 

It would be nice if you could test it on your side too:
- Upgrade to Dune 2.8
- Clean and rebuild your project
- Reload VSCode window

If you want to print the generated configuration you can use:
- From dune: `dune ocaml-merlin --dump-config=$(pwd)` from your source directory
- From merlin: `ocamlmerlin single dump-configuration -filename main.ml <main.ml | jq` where `main.ml` is the name of your source file.

With Dune 2.8.0 these configurations appear perfectly valid to me and Merlin gives correct answers when calling "type-enclosing" directly. @rgrinberg : I am under the impression that this is in fact an OCamlLSP issue, but I am unsure why.

**Comment by HavinLeung** [comment]
I upgraded to Dune 2.8 and cleaned/rebuilt, but now `dune build` doesn't generate the .merlin files for me at all

**Comment by voodoos** [comment]
That is actually a feature, starting with dune 2.8.0 `.merlin` files are no more promoted to the sources. That is why is suggested two commands to get the configuration in my previous comment, sorry if that wasn't clear.

Recent version of Merlin (>= 3.4.0) and OCamlLSP (>= 1.3.0) will query Dune directly for their configuration for each opened file (this allows dune to provide more precise information than the old `.merlin` files that were folder-scoped).

Your editor integration should not have stopped working with this new version, even if the `.merlin` files are missing.

**Comment by HavinLeung** [comment]
that's very odd - my versions were all sufficiently high yet both my spacemacs and vscode integration was broken. There's nothing special you need to do, don't even need a .merlin file?

I'll investigate more.

**Comment by voodoos** [comment]
No, there should be nothing more to do than build the project once with the new dune and refresh your buffers.

Dune must be in your path for it to work, don't forget to ```eval `(opam env)``` if you have multiple opam switches.

When you build your project with dune 2.8 merlin configuration files are written in a binary format in your `_build` folder. You can query this configuration from your source folders with the commands in my first comment.

If the issue persists, copy here the output of these two commands as it can help us to understand the problem.

**Comment by HavinLeung** [comment]
Not sure what was different this time around, maybe I was just tired. Reran `opam upgrade` and `opam user-setup install` and things worked fine in the project without space in filepath on both spacemacs and vscode. Broken in vscode only for the project with the space in filepath.

```
havinleung@Macbook_Pro a1_code_copy % dune --version
2.8.0
havinleung@Macbook_Pro a1_code_copy % ocamlc --version
4.07.1
havinleung@Macbook_Pro a1_code_copy % ocamllsp --version
1.4.0
```

[gist of some commands in ran in project without space in filepath](https://gist.github.com/HavinLeung/9e011c5818267b7f2842c8a49e63851f)

[gist of some commands in ran in project with the space in filepath](https://gist.github.com/HavinLeung/25a8c0adf74a3f033f2a85581cf44ee2)

**Comment by d4hines** [comment]
I believe I've uncovered a problem in the same class.

If the name of the folder at the root of my project has an `@` sign in it (concretely, `dh@table-of-values2`), OCaml Platform Extension in VS Code fails to report the type information of types from imported modules, as though it can't find any merlin information. If I rename the folder to remove the `@` sign (concretely, `foo`) and rebuild, everything works.

Version info:
```
[d4hines@RADAH foo]$ esy --version
0.6.10
[d4hines@RADAH foo]$ dune --version
2.8.4
[d4hines@RADAH foo]$ ocamlc --version
4.10.2
[d4hines@RADAH foo]$ ocamllsp --version
1.4.1
```

**Comment by rgrinberg** [comment]
@d4hines are you still able to reproduce this issue with the latest versions of ocamllsp and dune?

**Comment by d4hines** [comment]
@rgrinberg, I updated dune and ocaml-lsp-server to 3.1.1 and 1.11.14 respectively. I did the following:

```
git clone https://github.com/d4hines/hello-dream hello-dream@ocaml-lsp-test
cd hello-dream@ocaml-lsp-test
esy
esy dune build @check
code .
```
And set the sandbox to the local esy switch.

ocamllsp doesn't crash, but VS Code wasn't giving me proper type definitions, e.g. `Unbound module: Dream` when the same commands in a folder named `hello-dream` give proper type definitions.

**Comment by voodoos** [comment]
Thanks for the confirmation, I will have a look.

**Comment by voodoos** [comment]
I was able to reproduce with your project and using ocaml-lsp in vscode.
However I am not able to reproduce in Merlin's testsuite or by running Merlin manually:
```
❯ ocamlmerlin single errors -filename hello.ml <hello.ml | jq
{
  "class": "return",
  "value": [],
  "notifications": [],
  "timing": {
    "clock": 52,
    "cpu": 21,
    "query": 3,
    "pp": 0,
    "reader": 0,
    "ppx": 0,
    "typer": 17,
    "error": 0
  }
}
```

@rgrinberg could this be an ocaml-lsp specific issue ?
